File: 695MaxAreaofIsland/MaxAreaofIsland.py
# -*- coding: utf-8 -*-

from typing import List

class MaxAreaofIsland:
    def maxAreaofIsland(self, grid: "List[List[int]]") -> "int":
        m = len(grid)
        if not m: return 0
        n = len(grid[0])
        visited = [[False for _ in range(n)] for _ in range(m)]
        dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        ans = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1 and not visited[i][j]:
                    area = 1 ## found America! call me Columbus
                    q = [(i, j)]
                    visited[i][j] = True
                    ## bfs
                    while q:
                        x, y = q.pop(0)
                        for dx, dy in dirs:
                            nx, ny = x + dx, y + dy
                            if 0 <= nx < m and 0 <= ny < n and\
                                    not visited[nx][ny] and grid[nx][ny] == 1:
                                area += 1
                                visited[nx][ny] = True
                                q.append((nx, ny))
                    ans = max(ans, area)
        return ans


def main():
    test = MaxAreaofIsland()
    print(
        test.maxAreaofIsland(
            [
                [0, 1, 1, 0, 1],
                [1, 0, 1, 0, 1],
                [0, 1, 1, 0, 1],
                [0, 1, 0, 0, 1],
            ]
        )
    )

if __name__ == "__main__":
    main()

File: 695MaxAreaofIsland/MaxAreaofIsland.cpp


#include <vector>
#include <queue>
#include <iostream>
using std::vector;
using std::queue;
using std::pair;
using std::size_t;

class MaxAreaofIsland {
public:
    int maxAreaofIsland(vector<vector<int> > grid) {
        size_t m = grid.size();
        if (m == 0) {
            return 0;
        }
        size_t n = grid[0].size();
        vector<pair<int, int> > dirs;
        dirs.push_back(std::make_pair(1, 0));
        dirs.push_back(std::make_pair(-1, 0));
        dirs.push_back(std::make_pair(0, 1));
        dirs.push_back(std::make_pair(0, -1));
        vector<vector<bool> > visited(m, vector<bool>(n, false));
        // for (size_t i = 0; i < m; ++i) {
        //     for (size_t j = 0; j < n; ++j) {
        //         visited[i].push_back(false);
        //     }
        // }
        int ans = 0;
        for (size_t i = 0; i < m; ++i) {
            for (size_t j = 0; j < n; ++j) {
                if (!visited[i][j] && grid[i][j] == 1) {
                    // found new island
                    queue<pair<size_t, size_t> > q;
                    q.push(std::make_pair(i, j));
                    visited[i][j] = true;
                    size_t area = 1;
                    while (!q.empty()) {
                        pair<size_t, size_t> cur = q.front();
                        q.pop();
                        for (size_t k = 0; k < dirs.size(); ++k) {
                            pair<int, int> d = dirs[k];
                            int nx = cur.first + d.first;
                            int ny = cur.second + d.second;
                            if (
                                nx >= 0 && nx < (int)m &&
                                ny >= 0 && ny < (int)n &&
                                !visited[nx][ny] && grid[nx][ny] == 1
                            ) {
                                area++;
                                visited[nx][ny] = true;
                                q.push(std::make_pair(nx, ny));
                            }
                        }
                    }
                    ans = std::max(ans, (int)area);
                }
            }
        }
        return ans;
    }
};


int main() {
    MaxAreaofIsland test;
    vector<vector<int> > t1;
    t1.push_back(vector<int>());
    t1.push_back(vector<int>());
    t1[0].push_back(1);
    t1[0].push_back(0);
    t1[1].push_back(0);
    t1[1].push_back(0);
    std::cout << test.maxAreaofIsland(t1) << std::endl;
}
